### PR TITLE
add new user - more commits with modules

### PR DESCRIPTION
**Pull request by sshi100**
Together with https://github.com/commitdev/terraform-aws-zero/pull/6

### CODE REVIEW

File: templates/terraform/environments/prod/user_access.tf [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# define policy documents for developer policy
+data "aws_iam_policy_document" "developer_access" {
+  # EKS
+  statement {
+    effect = "Allow"
+    actions = [
+      "eks:List*",
+      "eks:Describe*"
+    ]
+    resources = ["arn:aws:eks:::cluster/piggycloud-me-stage-*"]
```
Comment (bmonkman) [inline templates/terraform/environments/prod/user_access.tf:10]: Shouldn't be "piggycloud"

Comment (sshi100) [reply templates/terraform/environments/prod/user_access.tf]: done

File: templates/terraform/README.md [hunk 2]
```diff
@@ -102,7 +102,15 @@
  make update-k8s-conf
  ```
 
+## User Access
 
+You may want to add people for AWS and Kubernetes access.
+ 1. Change user list with role and environment assignment in environments/shared/main.tf
+ 2. Change user access policy in environments/<env>/user_access.tf
+After applied, people can switch to assigned k8s context, in the root of the project, run:
```
Comment (bmonkman) [inline templates/terraform/README.md:110]: We should note that they will need to apply in shared, and in each environment they changed.

File: templates/Makefile [hunk 3]
```diff
@@ -40,7 +41,7 @@ apply-k8s-utils:
 	terraform apply $(AUTO_APPROVE)
```
Comment (bmonkman) [inline templates/Makefile:42]: We will need a new make target to apply the shared environment.

File: templates/terraform/environments/prod/user_access.tf [hunk 4]
```diff
@@ -0,0 +1,111 @@
+# define AWS policy documents for developer
+data "aws_iam_policy_document" "developer_access" {
+  # IAM
+  statement {
```
Comment (bmonkman) [inline templates/terraform/environments/prod/user_access.tf:4]: This should be moved to the module. Added a note there.

File: templates/terraform/environments/prod/user_access.tf [hunk 5]
```diff
@@ -0,0 +1,111 @@
+# define AWS policy documents for developer
+data "aws_iam_policy_document" "developer_access" {
```
Comment (bmonkman) [inline templates/terraform/environments/prod/user_access.tf:2]: At some point we will want to revisit these policies, add stuff like enforcing MFA, etc. but this should be fine for now.

File: templates/terraform/modules/environment/main.tf [hunk 6]
```diff
@@ -8,6 +8,31 @@ data "aws_iam_user" "ci_user" {
...
+}
+
+locals {
+  role_name_list = var.roles.*.name
+  users = [
+    for u in data.terraform_remote_state.shared.outputs.user_role_mapping : {
+      name  = u.name
+      roles = [
+        for r in u.roles :
+        r.name if contains(local.role_name_list, r.name) && contains(r.environments, var.environment)
```
Comment (bmonkman) [inline templates/terraform/modules/environment/main.tf:30]: Nice, much cleaner now!

File: templates/terraform/modules/environment/main.tf [hunk 7]
```diff
@@ -8,6 +8,31 @@ data "aws_iam_user" "ci_user" {
   user_name = "${var.project}-ci-user" # Should have been created in the bootstrap process
 }
 
+# get users from remote state of "shared"
+data "terraform_remote_state" "shared" {
```
Comment (bmonkman) [inline templates/terraform/modules/environment/main.tf:12]: Let's move this declaration out to the root and then pass the user role mapping in to the environment module? I don't like having the state reference hidden in a module (even if it is a local module..)

Comment (sshi100) [reply templates/terraform/modules/environment/main.tf:12]: move to the root and applied corresponding changes

File: templates/terraform/environments/prod/user_access.tf (hunk 4 above)
Comment (sshi100) [reply templates/terraform/environments/prod/user_access.tf:4]: done

File: templates/Makefile (hunk 3 above)
Comment (sshi100) [reply templates/Makefile:42]: done apply-shared-env prior to apply-env

File: templates/terraform/README.md (hunk 2 above)
Comment (sshi100) [reply templates/terraform/README.md:110]: done

File: templates/terraform/environments/prod/user_access.tf [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+      "ecr:DescribeRepositories"
+    ]
+    resources = ["*"]
+  }
+
+  # S3
+  statement {
+    effect    = "Allow"
+    actions   = ["s3:ListBucket"]
+    resources = ["arn:aws:s3:::*<% index .Params `productionHostRoot` %>"]
```
Comment (bmonkman) [inline templates/terraform/environments/prod/user_access.tf:29]: Hmm if we're going to be using these values a lot more at the root how about putting them into locals in `main.tf`?

File: templates/terraform/environments/shared/main.tf [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+# Instantiate the environment
+locals {
+  # Project configuration
+  project = "<% .Name %>"
+
+  # Users configuration
+  users = [
+    {
+      name  = "dev1"
```
Comment (bmonkman) [inline templates/terraform/environments/shared/main.tf:25]: Let's comment out these users, we don't actually want to create them.

File: templates/terraform/environments/stage/main.tf [hunk 10]
```diff
@@ -35,15 +54,15 @@ module "stage" {
   eks_worker_asg_max_size  = 3
 
   # EKS-Optimized AMI for your region: https://docs.aws.amazon.com/eks/latest/userguide/eks-optimized-ami.html
-  # https://<% index .Params `region` %>.console.aws.amazon.com/systems-manager/parameters/%252Faws%252Fservice%252Feks%252Foptimized-ami%252F1.17%252Famazon-linux-2%252Frecommended%252Fimage_id/description?region=<% index .Params `region` %>
+  # https://${local.region}.console.aws.amazon.com/systems-manager/parameters/%252Faws%252Fservice%252Feks%252Foptimized-ami%252F1.17%252Famazon-linux-2%252Frecommended%252Fimage_id/description?region=${local.region}
```
Comment (bmonkman) [inline templates/terraform/environments/stage/main.tf:57]: This has to be templated in, it's in a comment.

File: templates/terraform/modules/environment/main.tf [hunk 11]
```diff
@@ -28,7 +41,7 @@ data "aws_caller_identity" "current" {}
 # Provision the EKS cluster
 module "eks" {
   source  = "commitdev/zero/aws//modules/eks"
-  version = "0.0.2"
+  version = "0.1.2"
```
Comment (bmonkman) [inline templates/terraform/modules/environment/main.tf:44]: Could you please bump db and logging up to 0.1.2 as well?

**Review by bmonkman — APPROVED** [review]

**Pull request merged**